app/routes/course/assignments.ts

```diff
@@ -0,0 +1,120 @@
+import { INTERNAL_SERVER_ERROR, OK } from 'http-status-codes';
+import * as Router from 'koa-router';
+import {
+    IApiResponse,
+    AssignmentModel,
+    IAssignmentModel,
+    IUserSession,
+    IEventModel,
+    TaskModel,
+    AssignmentStatus,
+} from '../../models';
+
+type AssignmentResponse = {
+    _id: string;
+    assignmentRepo?: string;
+    author?: string;
+    checkDate?: number;
+    completeDate?: number;
+    courseId: string;
+    deadlineDate: number;
+    endDateTime?: number;
+    mentorComment?: string;
+    mentorId?: string;
+    score: number;
+    startDateTime: number;
+    status: string;
+    studentComment?: string;
+    studentId: string;
+    taskId: string;
+    taskType?: string;
+    title: string;
+    type: string;
+    urlToDescription?: string;
+    whoChecks?: string;
+};
+
+type INormalizeAssignments = {
+    isEndAssignment: boolean;
+    isActiveAssignment: boolean;
+    assignment: AssignmentResponse;
+};
+
+type NormalizeAssignmentsData = {
+    assignments: INormalizeAssignments[];
+};
+
+export const courseAssignmentsRoute = async (ctx: Router.IRouterContext) => {
+    try {
+        const userSession: IUserSession = ctx.state.user!;
+        const { id: courseId } = ctx.params;
+        const studentId: string = userSession._id;
+        const assignments: IAssignmentModel[] = await getAssignmentsByCourseIdAndUserId(courseId, studentId);
+        const arrayOfTasksId: string[] = [];
+        for (const index in assignments) {
+            if (assignments[index]) {
+                arrayOfTasksId.push(assignments[index].taskId);
+            }
+        }
+        const tasks: IEventModel[] = await getTasksByTaskId(arrayOfTasksId);
+        const combinedAssignmentsAndTasks: AssignmentResponse[] = [];
+        for (const index in assignments) {
+            if (assignments[index] && tasks[index]) {
+                combinedAssignmentsAndTasks[index] = { ...Object.create(null), ...tasks[index], ...assignments[index] };
+            }
```

Comment:
Please, rewrite this part. I think it would be better to use  `populate` or some aggregation here.
Docs: https://mongoosejs.com/docs/populate.html

Reply:
It is impossible to implement this part with *aggregation*,  because we have to compare taskId(String) and _id(ObjectId) and there is no way to convert them to the same type(an example is given below). Tried with : _id.str , { "$toString": "$_id" }, $convert. We can try to change the type of "taskId" in the Assignment collection to ObjectId, but I don't think it would be a good idea.
```javascript
const assignments = await AssignmentModel.aggregate([
{
$match: {courseId,studentId}
},
{
$lookup:{
from: 'tasks',
localField: 'taskId', // String
foreignField: '_id', // ObjectId
as 'task'
}
},
{
$unwind: '$task'
},
])
```
Perhaps this can be done with *populate*, but so far I have not found a way to implement it.